Integration tests fail with Matrix Connection aborted.

## Problem Definition

In Travis some integration tests seem to be continuously failing in a flaky way with various types of connection aborted messages.

[Here](https://travis-ci.org/raiden-network/raiden/jobs/415129600) is an example run.

And here is the traceback:

```
==================================== ERRORS ====================================
_ ERROR at setup of test_settled_lock[transport_config0-channels_per_node0-2] __
blockchain_p2p_ports = [29944]
blockchain_private_keys = [b'\xf6\x89R\xf0\x8b`\xb7\x9a]\xbd>\xc7T\xee3\xe6\xe5!\x01\xfc\xf1\xd8\x05_\xc8\x99y\xc2S\xb1H2']
blockchain_rpc_ports = [29943], blockchain_type = 'geth'
deploy_key = b'\xc5\x01t\xf9\xf7\xc8A\x8fo^\xafp\xb0\x91S&I0bh\xf2\xfcs\xce\xcbW\xa0\xdb\xf6\x94\xf6\x89'
private_keys = [b'k\xfd\xc8\xf1\xd4R{D\xab\xd0\x9f\xda\x8eQ\xe44\x9e\xe2\xc72\xb3\xb3\n\xb7\xc9`\xb8\x06\x83\x89\xc4\xc0', b'\x15\x86\xe9\xd2Y\x05H\xd6=\x91\xbb\xe0\xc0d\x06w\x825\xe7\xd6\xf6\x80\x06\x95\xa4<\t\x98\x88K\xb2\x84']
random_marker = '791cbf60bb62cf5f242f92671'
request = <SubRequest 'web3' for <Function 'test_settled_lock[transport_config0-channels_per_node0-2]'>>
tmpdir = local('/tmp/pytest-of-travis/pytest-0/test_settled_lock_transport_co0')
ethereum_tester = <eth_tester.main.EthereumTester object at 0x7f6c8c0b48d0>
chain_id = 337
    @pytest.fixture
    def web3(
            blockchain_p2p_ports,
            blockchain_private_keys,
            blockchain_rpc_ports,
            blockchain_type,
            deploy_key,
            private_keys,
            random_marker,
            request,
            tmpdir,
            ethereum_tester,
            chain_id,
    ):
        """ Starts a private chain with accounts funded. """
        # include the deploy key in the list of funded accounts
        keys_to_fund = set(private_keys)
        keys_to_fund.add(deploy_key)
        keys_to_fund = sorted(keys_to_fund)
    
        if blockchain_type == 'geth':
            host = '0.0.0.0'
            rpc_port = blockchain_rpc_ports[0]
            endpoint = f'http://{host}:{rpc_port}'
            web3 = Web3(HTTPProvider(endpoint))
    
            assert len(blockchain_private_keys) == len(blockchain_rpc_ports)
            assert len(blockchain_private_keys) == len(blockchain_p2p_ports)
    
            geth_nodes = [
                GethNodeDescription(
                    key,
                    rpc,
                    p2p,
                    miner=(pos == 0),
                )
                for pos, (key, rpc, p2p) in enumerate(zip(
                    blockchain_private_keys,
                    blockchain_rpc_ports,
                    blockchain_p2p_ports,
                ))
            ]
    
            accounts_to_fund = [
                privatekey_to_address(key)
                for key in keys_to_fund
            ]
    
            geth_processes = geth_run_private_blockchain(
                web3,
                accounts_to_fund,
                geth_nodes,
                str(tmpdir),
                chain_id,
                request.config.option.verbose,
>               random_marker,
            )
accounts_to_fund = [b'\xba\x9d\xd8\x8d\x16\xe9\xc3$\x0b\xcb\xfd\n\xfa\xfaO+\xe2xCp', b'\xce;\x04\x96\xca.(\xd2\xbb\xaf(\x10|y\xc4\xb54\xb7\x95\xc9', b' \xd8\n\x83\x9b.\xade\xf2qr-\x073\x90w\xf2\x14\x04\xdb']
blockchain_p2p_ports = [29944]
blockchain_private_keys = [b'\xf6\x89R\xf0\x8b`\xb7\x9a]\xbd>\xc7T\xee3\xe6\xe5!\x01\xfc\xf1\xd8\x05_\xc8\x99y\xc2S\xb1H2']
blockchain_rpc_ports = [29943]
blockchain_type = 'geth'
chain_id   = 337
deploy_key = b'\xc5\x01t\xf9\xf7\xc8A\x8fo^\xafp\xb0\x91S&I0bh\xf2\xfcs\xce\xcbW\xa0\xdb\xf6\x94\xf6\x89'
endpoint   = 'http://0.0.0.0:29943'
ethereum_tester = <eth_tester.main.EthereumTester object at 0x7f6c8c0b48d0>
geth_nodes = [GethNodeDescription(private_key=b'\xf6\x89R\xf0\x8b`\xb7\x9a]\xbd>\xc7T\xee3\xe6\xe5!\x01\xfc\xf1\xd8\x05_\xc8\x99y\xc2S\xb1H2', rpc_port=29943, p2p_port=29944, miner=True)]
host       = '0.0.0.0'
keys_to_fund = [b'\x15\x86\xe9\xd2Y\x05H\xd6=\x91\xbb\xe0\xc0d\x06w\x825\xe7\xd6\xf6\x80\x06\x95\xa4<\t\x98\x88K\xb2\x84', b'k\xfd\xc...xb8\x06\x83\x89\xc4\xc0', b'\xc5\x01t\xf9\xf7\xc8A\x8fo^\xafp\xb0\x91S&I0bh\xf2\xfcs\xce\xcbW\xa0\xdb\xf6\x94\xf6\x89']
private_keys = [b'k\xfd\xc8\xf1\xd4R{D\xab\xd0\x9f\xda\x8eQ\xe44\x9e\xe2\xc72\xb3\xb3\n\xb7\xc9`\xb8\x06\x83\x89\xc4\xc0', b'\x15\x86\xe9\xd2Y\x05H\xd6=\x91\xbb\xe0\xc0d\x06w\x825\xe7\xd6\xf6\x80\x06\x95\xa4<\t\x98\x88K\xb2\x84']
random_marker = '791cbf60bb62cf5f242f92671'
request    = <SubRequest 'web3' for <Function 'test_settled_lock[transport_config0-channels_per_node0-2]'>>
rpc_port   = 29943
tmpdir     = local('/tmp/pytest-of-travis/pytest-0/test_settled_lock_transport_co0')
web3       = <web3.main.Web3 object at 0x7f6c7572eb38>
raiden/tests/integration/fixtures/blockchain.py:97: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
raiden/tests/utils/geth.py:437: in geth_run_private_blockchain
    logdir,
raiden/tests/utils/geth.py:336: in geth_run_nodes
    verbosity,
raiden/tests/utils/geth.py:304: in geth_nodes_to_cmds
    geth_prepare_datadir(datadir, genesis_file)
raiden/tests/utils/geth.py:290: in geth_prepare_datadir
    geth_init_datadir(datadir, node_genesis_path)
raiden/tests/utils/geth.py:200: in geth_init_datadir
    subprocess.check_output(args, stderr=subprocess.STDOUT)
../../../virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/subprocess.py:325: in check_output
    with Popen(*popenargs, stdout=PIPE, **kwargs) as process:
../../../virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/subprocess.py:604: in __init__
    restore_signals, start_new_session)
../../../virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/subprocess.py:1412: in _execute_child
    data = errpipe_read.read()
../../../virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/_fileobjectposix.py:113: in readall
    data = self.__read(DEFAULT_BUFFER_SIZE)
../../../virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/_fileobjectposix.py:108: in __read
    self.hub.wait(self._read_event)
src/gevent/_hub_primitives.py:46: in gevent.__hub_primitives.WaitOperationsGreenlet.wait
    ???
src/gevent/_hub_primitives.py:55: in gevent.__hub_primitives.WaitOperationsGreenlet.wait
    ???
src/gevent/_waiter.py:152: in gevent.__waiter.Waiter.get
    ???
src/gevent/_greenlet_primitives.py:59: in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
    ???
src/gevent/_greenlet_primitives.py:59: in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
    ???
src/gevent/_greenlet_primitives.py:63: in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>   ???
E   raiden.exceptions.TransportError: Unexpected error while communicating with Matrix homeserver: Something went wrong in GET requesting http://localhost:8008/_matrix/client/r0/sync: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))
SwitchOutGreenletWithLoop = <class 'gevent.__greenlet_primitives.SwitchOutGreenletWithLoop'>
TrackedRawGreenlet = <class 'gevent.__greenlet_primitives.TrackedRawGreenlet'>
__all__    = ['TrackedRawGreenlet', 'SwitchOutGreenletWithLoop']
__builtins__ = <builtins>
__doc__    = '\nA collection of primitives used by the hub, and suitable for\ncompilation with Cython because of their frequency of use.\n\n'
__file__   = '/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/__greenlet_primitives.cpython-36m-x86_64-linux-gnu.so'
__loader__ = <_frozen_importlib_external.ExtensionFileLoader object at 0x7f6c8fd1cf98>
__name__   = 'gevent.__greenlet_primitives'
__package__ = 'gevent'
__pyx_capi__ = {'BlockingSwitchOutError': <capsule object "PyObject *" at 0x7f6c8fd04a50>, '_greenlet_imported': <capsule object "int" at 0x7f6c8fd04a80>, 'wref': <capsule object "PyObject *" at 0x7f6c8fd049f0>}
__spec__   = ModuleSpec(name='gevent.__greenlet_primitives', loader=<_frozen_importlib_external.ExtensionFileLoader object at 0x7f6...ravis/virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/__greenlet_primitives.cpython-36m-x86_64-linux-gnu.so')
__test__   = {}
_greenlet_switch = <method 'switch' of 'greenlet.greenlet' objects>
_init      = <cyfunction _init at 0x7f6c91081f60>
getcurrent = <built-in function getcurrent>
greenlet_init = <cyfunction <lambda> at 0x7f6c91081d38>
import_c_accel = <function import_c_accel at 0x7f6c8fcad488>
src/gevent/__greenlet_primitives.pxd:35: TransportError
```

--- a/raiden/network/transport/matrix.py
+++ b/raiden/network/transport/matrix.py
@@ -270,11 +270,11 @@ class MatrixTransport:
         # cause pending retries to be aborted
         for async_result in self._messageids_to_asyncresult.values():
             async_result.set(False)
-        gevent.wait(self.greenlets, timeout=self._logout_timeout)
 
         try:
             self._client.set_presence_state(UserPresence.OFFLINE.value)
             with gevent.Timeout(self._logout_timeout, 'Logging out despite unjoined greenlets.'):
+                gevent.wait(self.greenlets, timeout=self._logout_timeout / 2)
                 self._client.stop_listener_thread()
         except gevent.Timeout as er:
             self.log.warning('Matrix stop timeout', _error=er)

--- a/raiden/network/transport/matrix.py
+++ b/raiden/network/transport/matrix.py
@@ -201,7 +201,6 @@ class MatrixTransport:
         # TODO: Add (better) error handling strategy
         self._client.start_listener_thread()
         self._client.sync_thread.link_exception_safe(self._client_exception_handler)
-        self.greenlets.append(self._client.sync_thread)
 
         # TODO: Add greenlet that regularly refreshes our presence state
         self._client.set_presence_state(UserPresence.ONLINE.value)
@@ -266,19 +265,22 @@ class MatrixTransport:
         if not self._running:
             return
         self._running = False
-        self._client.set_presence_state(UserPresence.OFFLINE.value)
-        self._client.stop_listener_thread()
-        self._client.sync_thread = None
 
         # Set all the pending results to False, this will also
         # cause pending retries to be aborted
         for async_result in self._messageids_to_asyncresult.values():
             async_result.set(False)
+        gevent.wait(self.greenlets, timeout=self._logout_timeout)
 
         try:
-            self._client.join_and_logout(self.greenlets, timeout=self._logout_timeout)
-        except RuntimeError as error:
-            self.log.critical(str(error))
+            self._client.set_presence_state(UserPresence.OFFLINE.value)
+            with gevent.Timeout(self._logout_timeout, 'Logging out despite unjoined greenlets.'):
+                self._client.stop_listener_thread()
+        except gevent.Timeout as er:
+            self.log.warning('Matrix stop timeout', _error=er)
+        finally:
+            self._client.greenlets.clear()
+            self._client.logout()
 
     @property
     def _user_id(self) -> Optional[str]:

--- a/raiden/tests/integration/fixtures/transport.py
+++ b/raiden/tests/integration/fixtures/transport.py
@@ -32,7 +32,6 @@ def skip_if_not_udp(request):
 
 @pytest.fixture
 def local_matrix_server(transport_config):
-
     if not transport_config.protocol == TransportProtocol.MATRIX:
         yield None
         return
@@ -45,14 +44,11 @@ def local_matrix_server(transport_config):
         return
 
     # otherwise, run our own local server
-    matrix = HTTPExecutor(
+    with HTTPExecutor(
         transport_config.parameters.command,
         url=urljoin(server, '/_matrix/client/versions'),
         method='GET',
         timeout=30,
         shell=True,
-    )
-
-    matrix.start()
-    yield server
-    matrix.stop()
+    ):
+        yield server

--- a/requirements.txt
+++ b/requirements.txt
@@ -16,7 +16,7 @@ netifaces==0.10.7
 networkx==2.1.0
 psutil==5.4.5
 pycryptodome==3.6.1
-raiden-libs==0.1.3
+git+https://github.com/raiden-network/raiden-libs.git@3b0c83f99c6f5994d3db74fe78ded0feb13f1350#egg=raiden-libs
 raiden-contracts==0.2.0
 webargs==1.8.1
 eth-keyfile==0.5.1

--- a/setup.py
+++ b/setup.py
@@ -81,6 +81,11 @@ history = ''
 
 install_requires_replacements = {
     'git+https://github.com/LefterisJP/pystun@develop#egg=pystun': 'pystun',
+    (
+        'git+https://github.com/raiden-network/raiden-libs.git'
+        '@3b0c83f99c6f5994d3db74fe78ded0feb13f1350'
+        '#egg=raiden-libs'
+    ): 'raiden-libs',
 }
 
 install_requires = list(set(

